### PR TITLE
feat: printer

### PR DESCRIPTION
**Pull request by tyler58546**
## Summary of Changes

* The printer can now be used.
    * Once a page has been printed, clicking on it will move it next to the MCDU
    * If multiple pages have been printed, you can switch between them ~using the arrows~ by clicking on the sides of the paper
    * Pages can be discarded by clicking on the bottom of the page

## Screenshots (if necessary)

image
image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

Request ATIS or weather from the MCDU and press the print button. Make sure the printer behaves as described.

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

File: A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                displayedPage = pagesPrinted - 1;
+                SimVar.SetSimVarValue("L:A32NX_PRINT_PAGE_OFFSET", "number", 0);
+            }
+        }
+
+        let newLines = '';
+        for (const value of (pages[displayedPage] || [])) {
+            newLines += `<span class="line">${value}<br/></span>`;
+        }
+        this.lines.innerHTML = newLines;
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.js:81]: This should not run every frame. Can you make it not update the DOM if nothing changed ?

File: src/behavior/src/A32NX_Interior_Printer.xml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    </Template>
+    <Template Name="FBW_Airbus_PAPER_PAGE_BUTTON">
+        <DefaultTemplateParameters>
+            <NODE_ID>#NODE_ID#</NODE_ID>
+            <CURSOR>Grab</CURSOR>
+            <DIRECTION>1</DIRECTION>
+            <PAGE_OFFSET_VAR>L:A32NX_PRINT_PAGE_OFFSET</PAGE_OFFSET_VAR>
+        </DefaultTemplateParameters>
+        <Component ID="#NODE_ID#" Node="#NODE_ID#">
+            <UseTemplate Name = "ASOBO_GT_Interaction_LeftSingle_Leave_Code">
```
Comment (Benjozork) [inline src/behavior/src/A32NX_Interior_Printer.xml:67]:
```suggestion
            <UseTemplate Name="ASOBO_GT_Interaction_LeftSingle_Leave_Code">
```

**Review by Benjozork — CHANGES_REQUESTED** [review]

**Review by Benjozork — APPROVED** [review]
LGTM

**Comment by wpine215** [comment]
Category: QA Tester/Trial
Name: Iceman#5540
Date of testing: 1/22/2021
Version of the sim: 1.12.13.0
PR Tested: #3135
Tier of testing: 1/2/3
Changes to observe:
Test weather and ATIS printouts.

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Loaded sim on runway.
2. Retrieved KORD METAR
3. Retrieved KORD ATIS
4. Tested switching between papers.

Media: 
From proper captain seat position, left arrow is barely visible (only a few pixels):
image

Suggestion: Perhaps a page corner texture can be used in place of arrows for back/forward? Would look more realistic imo.
image

Also, would be good to have printout sounds and page turning sounds in a separate PR by Boris.

Issues (If Any): None

Overall Rating: Good.
Conclusions of the testing: Works as described.

**Review by devsnek — DISMISSED** [review]
marking requested changes until this is moved into src/instruments

**Comment by marinerhally** [comment]
Category: QA Tester
Name: Mariner Hally#0715
Date of testing: 23/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3135
Tier of testing: 2
Changes to observe:
The printer can now be used.
Once a page has been printed, clicking on it will move it next to the MCDU.
If multiple pages have been printed, you can switch between them using the arrows.


Testing technique: Followed instructions.
Steps you've made to test the Pull Request:
1. Spawned C/D at EGKK (VATSIM active at the time)
2. Requested METAR for EGKK and EGLL
3. Requested ATIS for EGKK and EGCC
4. Switched between pages

Flight notes: None

Media: METAR request for EGKK and EGLL. (see below)
Printer


Issues (If Any): None

Overall Rating: Good
Conclusions of the testing: When requesting 2 METAR together not sure irl but would there be a space between them or even if the second airport starting on the next line. If you see from the picture EGLL comes straight after EGKK at first I personally thought it had only printed one until closer inspection just think it would look more professional spaced or on the next line.

**Comment by tyler58546** [comment]
Thank you all for the feedback. I have made a few changes:
* You can now switch pages by clicking anywhere on the left or right side of the paper
* The paper has been moved to the center of the pedestal
* An arrow cursor is now used when hovering over the paper
* The arrow icons have been removed

> Conclusions of the testing: When requesting 2 METAR together not sure irl but would there be a space between them or even if the second airport starting on the next line. If you see from the picture EGLL comes straight after EGKK at first I personally thought it had only printed one until closer inspection just think it would look more professional spaced or on the next line.

I will look into this.

**Comment by ErickSharp** [comment]
Category: QA Tester
Name: Erick [Z-6]#6484
Date of testing: Current date 23/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3135 
Object of test: Added printer functionality

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. Get loaded in cold and dark at LFPG (Paris)
2. Get cockpit set up and input flight plan data
3. Go to the WX Request menu and request METAR for both ICAO's
4. Wait for company message indication on ECAM to display
5. Navigate to the Received Messages menu in MCDU, click "Print"
6. See METAR print out from the printer

Flight notes (If required): LFPG - EGLL

Issues (If Any): When printing out the METAR for two or more ICAO's the data of the two is merged, meaning there is no separation indicating what is from one ICAO and what is from another. 
Overall Rating: Good, minor improvement needed.
Conclusions of the testing: The printer works very well and is reliable. The only thing that is needed is a separation of METAR data from several ICAOs, as well as a way to scrap print-outs.

**Comment by stescouse** [comment]
Category: QA Tester
Name: Discord User stescouse#7271
Date of testing: 23.1.2021
Version of the sim: 1.12.13.0
PR Tested: #3135
Tier of testing: 2
Changes to observe: Request ATIS or weather from the MCDU and press the print button. Make sure the printer behaves as described.

Testing technique: TryOut
Steps you've made to test the Pull Request:

1. Spawned C&D at LDDU
2. Requested METAR for Weather at LDDU and LFPG.
3. Printed out METAR

print 1

4. Requested TAF for Weather at LDDU and LFPG.
5. Printed out TAF.

print 2

6  Checked page forward and page back for printouts.
7. Requested METAR on APPR to LFPG and printed.

print 3



Flight notes (If required: Full flight from LDDU to LFPG.

Issues (If Any): None


Overall Rating: Good
Conclusions of the testing: Great addition that saves switching around in MCDU to fill in APPR page.

**Comment by ChruutvoLuzi** [comment]
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 23.1.2021
Version of the sim: 1.12.13.0
PR Tested: #3135
Tier of testing: 2
Changes to observe:
Printer Printouts and Pages

Testing technique:TryOut at gate
Steps you've made to test the Pull:
1. Spawned at the Gate A44 in Zurich
2. Started the A/C
3. Tried Printing out METAR, TAFs and ATIS of LSZH and LSZB
4. Tried switching between the pages

Media:
image
image
Issues (If Any): None

Overall Rating: Good.
Conclusions of the testing: Works as described, however a few suggestions: As already mentioned (and you can see in Image 2) Data from two Airports get merged instead of being seperated (maybe by a line like there already is). Also it would be nice to remove printed Pages maybe by klicking an X in the bottom middle. Makes keeping things simpler. Otherwise very nice Feature

**Comment by liebbe10** [comment]
Wow that's so amazing! Thats really cool commitment to details. One Idea (besides this being absolutely amazing on its own..), would it be possible to "store" the papers in between the throttles and also trash them completely?
As said, its already amazing an just a fun low priority idea :D

**Comment by HugoTro** [comment]
Category: QA Tester
Name: Hugo #3307
Date of testing: Current date 23/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3135 
Tier of testing: 2
Changes to observe: Printer

Testing technique: Look-at
Steps you've made to test the Pull Request:
1. Requested TAF & METAR for LFRS and KLAX
2. Waited response for the 4 messages
3. Printed all of them

Issues: No issues observed

Overall Rating: Good
Conclusions of the testing: Works very good! If there is one thing to change in this PR it would be the page transitions, which look weird, when you turn a page it should not have any letter animation and just be instant

**Comment by BritishAvgeek** [comment]
Category: QA Tester
Name: BritishAvgeek#1436
Date of testing: 23/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3135 
Tier of testing: 2
Changes to observe:
The new Print functionality

Testing technique: Full Flight on VATSIM, livestreamed on my YT channel - Miami to Toncontin
Steps you've made to test the Pull Request:
1. Spawned C&D at KMIA, imported simbrief flight plan... pulled METAR/Wx Info via MCDU and printed it via printer
2. During flight, requested weather multiple times throughout the journey
3. Used FreeText between other aircraft and printed those messages out too
4. Pulled Wx Data onto the printer nearer arrival at Toncontin
5. Flicked between all the pages numerous times

Flight notes (When done a full flight):
Route: MNATE2 MNATE Y442 FUNDI UM335 ALVEK UM328 MANUR/N0457F360 DCT KARET DCT RNAV North Runway 02 via MELKO

Media: Livestream here: https://www.youtube.com/watch?v=eHJRoIdU0cw

Issues (If Any): When I received multiple messages fairly quickly from Print Request 13/14 it stopped printing for a short while. When I clicked print again, it jumped it straight to the "notepad area"... then I left it for 5mins or so, requested some Wx Data and that printed fine again... used the print function about 20-25 times during the flight and everything was great except for Print Request 13/14/15.
Steps to achieve the issue - As Described above

Overall Rating: Good but note the issue above
Conclusions of the testing: The overall functionality and the ability to print is awesome. Personally I don't mind the way the prints fade when switching and those watching the stream loved the feature too, though it does look a little weird when the top page dissolves into the new page!

The issues mentioned above seem to be because I was basically abusing the printer and seems as though it broke temporarily.

**Comment by cianwhite** [comment]
Category: QA Tester
Name: Cian#3596
Date of testing: 23/01/2021
Version of the sim: 1.12.13.0
PR Tested: #3135
Tier of testing: 2
Changes to observe: Printer functionality in requesting ATIS and METAR reports.

Testing technique: Testing the print functionality on the ground (C&D), at cruise level and from loading in for a mid-air start.
Steps you've made to test the Pull Request:
1. Started C&D at KMIA, started as normal and tried to request the ATIS and METAR for KMIA and KATL
2. The printouts successfully printed and I could interchange between them, although I found it hard to distinguish between the same METAR and ATIS reports (there doesn't seem to be an animation for the same report)
3. Took off and reached cruise level, requested another ATIS and METAR and all worked as expected
4. Spawned midair on an approach to KATL and could still print ATIS and METAR reports.

Route: HEDLY2 HEDLY DCT STYMY Q77 MJAMS/N0451F370 Q77 SHRKS/N0456F360 DCT LAIRI DCT LARZZ SITTH2

Media: https://streamable.com/3caawf

Issues (If Any): When I printed the same report, it can be hard to distinguish if you're actually changing pages. When you change between a different report it displays an animation. In this case I had 2 KMIA-KATL metars and 3 KATL ATIS printed out. My idea would be to either completely remove the animation or incorporate it for the same reports.

Steps to achieve the issue - Print the same report and you can see that there was no animation for the same report (see streamable). Nothing too serious but until I realised this I thought the page switching was broken. 

Overall Rating: Good just a small nitpick
Conclusions of the testing: The print function works really well, and is a very cool and interactive feature that adds to the overall quality of the mod. Think it will definitely be useful!
.

**Comment by tyler58546** [comment]
Thanks for testing everyone! I've made a few more changes based on your feedback:
* Fixed formatting when printing METARs for multiple airports
* Pages can now be discarded by clicking the bottom of the page

File: A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.js [hunk 3]
```diff
@@ -66,6 +68,16 @@ class LiveryPrinter extends TemplateElement {
                 displayedPage = pagesPrinted - 1;
                 SimVar.SetSimVarValue("L:A32NX_PRINT_PAGE_OFFSET", "number", 0);
             }
+
+            if (discard) {
+                pages.splice(displayedPage, 1);
+                localStorage.setItem("pages", JSON.stringify(pages));
+                pagesPrinted--;
+                SimVar.SetSimVarValue("L:A32NX_PAGES_PRINTED", "number", pagesPrinted);
+                SimVar.SetSimVarValue("L:A32NX_PAGES_PRINTED", "number", pagesPrinted);
```
Comment (Khitiara) [inline A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.js:77]: duplicate code?

**Comment by Awemeter** [comment]
Category: QA Tester
Name: Awemeter#7919 
Date of testing: 1/23/2021
Version of the sim: 1.12.13.0
PR Tested: #3135 
Tier of testing: 1
Changes to observe:
The printer can now be used.
Once a page has been printed, clicking on it will move it next to the MCDU
If multiple pages have been printed, you can switch between them using the arrows by clicking on the sides of the paper
Pages can be discarded by clicking on the bottom of the page

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request: Tried out
1. Spawned C&D at KORD 
2. Requested Vatsim atis from KORD and KMEM then printed it
3. Requested Metar from KDCA KATL KORD

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.




Issues (If Any): 
When you switch pages the old one fades into the new one.
Steps to achieve the issue:
Switch pages on the paper
Overall Rating: Good
Conclusions of the testing: Great feature nice work but I do think we should have the sound pr for the printer be merged before we merge this one.

**Comment by FlightSimmer7700** [comment]
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 24/1/21
Version of the sim: Versi1.12.13.0
PR Tested: #3135 
Tier of testing: 1
Changes to observe:
Printer added functionality

Testing technique: Loaded in and printed a few things
Steps you've made to test the Pull Request:
1. Loaded in
2. printed metars and TAF's
3. went to other printable pages

Flight notes (When done a full flight):

Media: Insert screenshots, recordings, or anything relevant to the review here.

In the CFDS page, the print button is not greyed out but also doesn't work
image

Same thing happens in the Avionics status page 
image

When trying to print something from a Free Text which is too long it just cuts it there and moves to the next message 
image

When something too long for the paper gets printed it just cuts it
image

Issues (If Any): 

- As described above the CFDS and Avionics status page are both not greyed out when they can't be utilized. 

- When printing something which doesn't fit on the page it just cuts off

Steps to achieve the issue: Try printing something too long and going to the pages mentioned

Overall Rating: To be improved
Conclusions of the testing: A solution for the overflow will be needed and the cyan print buttons which don't work yet should be greyed out

File: A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.html [hunk 4]
```diff
@@ -1,4 +1,4 @@
-<link rel="stylesheet" href="Printer.css" />
+<link rel="stylesheet" href="Printer.css?a=baaaaaaaa" />
```
Comment (Benjozork) [inline A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.html:1]: oops

Comment (tyler58546) [reply A32NX/html_ui/Pages/VLivery/Liveries/Printer/Printer.html]: oops

**Comment by FlyingCarpet7** [comment]
Awesome feature implementing the printer! Well done!
If I understand it correctly you are already aware of that the printer speed is animated a bit to fast? Should be maybe half of the current speed.

**Comment by ErikNL80** [comment]
I just raised an issue with the new printer functionality. Paper stays white. Or do I have to replace a cartridge :)

**Comment by andrejetcat** [comment]
I also recognise an minor bug,  I tested the printer once with an ATIS, and secound time with 3 METAR´s

The ATIS is in a proper font-size, and the 3 METAR´s appear in a very smaller font size.

**Pull request merged**